inst/include/concaveman/api.hpp

```diff
@@ -1,29 +1,107 @@
+#ifndef R_CONCAVEMAN_H
+#define R_CONCAVEMAN_H
+
 #pragma once
 
+#include <RcppCommon.h>
+#include <array>
+
+namespace Rcpp {
+namespace traits {
+
+//  for using Rcpp::as< >();
+template < typename T > class Exporter< std::array< T, 2 > >;
+template < typename T > class Exporter< std::vector< std::array< T, 2 > > >;
+
+} // traits
+} // Rcpp
+
+
 #include "Rcpp.h"
 
-namespace api {
+namespace Rcpp {
+namespace traits {
 
-Rcpp::DataFrame rcpp_concaveman_mat (Rcpp::NumericMatrix xy, Rcpp::IntegerVector hull_in,
-        const double concavity, const double length_threshold)
-{
-    const size_t num_points = static_cast <int> (xy.nrow ());
+    template < typename T > class Exporter< std::array< T, 2 > > {
+        typedef typename std::array< T, 2 > Point;
+        
+        const static int RTYPE = Rcpp::traits::r_sexptype_traits< T >::rtype;
+        Rcpp::Vector< RTYPE > vec;
+        
+    public:
+        Exporter( SEXP x ) : vec( x ) {
+            if( TYPEOF( x ) != RTYPE ) {
+                throw std::invalid_argument("decido - invalid R object for creating a Point");
+            }
+        }
+        
+        Point get() {
+            if( vec.length() != 2 ) {
+                Rcpp::stop("decido - each point in the polygon must have exactly two coordinates");
+            }
+            Point x({ vec[0], vec[1] });
+            return x;
+        }
+        
+    };
     
-    typedef double T;
-    typedef std::array <T, 2> point_type;
+    template < typename T > class Exporter< std::vector< std::array< T, 2 > > > {
+        typedef typename std::vector< std::array< T, 2 > > Polygon;
+        
+        const static int RTYPE = Rcpp::traits::r_sexptype_traits< T >::rtype;
+        Rcpp::Matrix< RTYPE > mat;
+        
+    public:
+        Exporter( SEXP x ) : mat( x ) {
+            if( TYPEOF( x ) != RTYPE ) {
+                throw std::invalid_argument("decido - invalid R object for creating a Polygon");
+            }
+        }
+        
+        Polygon get() {
+            R_xlen_t n_row = mat.nrow();
+            Polygon x( n_row );
+            R_xlen_t i;
+            for( i = 0; i < n_row; ++i ) {
+                Rcpp::Vector< RTYPE > v = mat( i, Rcpp::_);
+                x[i] = Rcpp::as< std::array< T, 2 > >( v );
+            }
+            return x;
+        }
+    };
 
-    std::vector <point_type> points (num_points);
-    for (auto i = 0; i < num_points; i++) {
-        points[i] = { xy (i, 0), xy (i, 1) };
-    }
+} // traits
+} // Rcpp
+
+#include "concaveman.h"
+
+typedef double T;   // TODO: does this NEED to be double? or can it follow RTYPE ?
+typedef std::array <T, 2> Point;
+typedef std::vector< Point > Points;
```

Comment:
No, that can't be `double`, and has to follow `RTYPE`, so those lines need to be:
```cpp
template <typename T>
using Point = std::array <T, 2>;

template <typename T>
using Points = std::vector <Point <T> >;
```
... then onto comment at line#85 below